test/e2e_node/services/server.go

```diff
@@ -282,7 +282,7 @@ func (s *server) kill() error {
 	}
 
 	// Attempt to shut down the process in a friendly manner before forcing it.
-	waitChan := make(chan error)
+	waitChan := make(chan error, 1)
```

Comment:
I think the goroutine needs to use a cancelable version of Wait(), if it exists, otherwise it has to poll.

The problem with this code isn't the lack of buffer, it's that the timeout case exits without cleaning things up. Adding a buffer sometimes masks that kind of problem but it's vastly preferable to actually fix it. Otherwise, the failure case has just been made more obscure and harder to debug.

In this case we can't say from looking at the code if the timeout was just too short, or if the kill command is not successful (e.g. we lack perms).

Reply:
@lavalamp thanks for the explanation. Couple of questions to understand this better.
1. cancelable version of Wait() : Use https://golang.org/pkg/context/#WithCancel ?
2. otherwise it has to poll.: For example, in case of timeout, we spawn another goroutine in the background to read on that channel and then log the error(s) appropriately ?

Reply:
1. Yes, Wait that takes a context, if it exists
2. Wait without a timeout can't work in this context, you'd have to find a version of the function that checks on status and returns immediately.

Reply:
@lavalamp , I came across this way of doing it. Create a new context with a deadline(timeout) and call os.exec with that context.  reference: https://golang.org/pkg/os/exec/#CommandContext

Here's a sample example where I am trying to kill the init process (as a non-superuser) once with -TERM and then with -KILL. To simulate their failure (w.r.t timeout), I use a sleep 5 command. For some reason this doesn't work in the play.golang. Hence pasting the entire snippet here.

```
package main

import (
    "context"
    "fmt"
    "os/exec"
    "time"
)

func main() {
    const timeout = 3 * time.Second
    for _, signal := range []string{"-TERM", "-KILL"} {
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel() // The cancel should be deferred so resources are cleaned up
                fmt.Printf("Killing process 1 with signal %s\n", signal)
		//cmd := exec.CommandContext(ctx, "kill", signal, "1") // Real command
		cmd := exec.CommandContext(ctx, "sleep", "5")        // Simulation command
                _, err := cmd.Output()

		if ctx.Err() == context.DeadlineExceeded {
			fmt.Printf("Kill timed out with signal: %s\n", signal)
			continue
		}

               if err != nil {
                       fmt.Printf("Error signaling process with %s: %v\n", signal, err)
              }
        }
}

```

Reply:
Right, that looks more correct.

Reply:
Changes done. Thank you.